Require that someone validate public_name correctly. (#456)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

the server uses wildcard names or added new names to the anonymity set.

public_name
: The DNS name of the client-facing server, i.e., the entity trusted
to update the ECH configuration. This is used to correct misconfigured clients,
as described in {{handle-server-response}}. This value MUST NOT begin or end
with an ASCII dot and MUST be parsable as a dot-separated sequence of LDH
labels, as defined in {{!RFC5890, Section 2.3.1}}. Clients MUST ignore any
`ECHConfig` structure whose public_name does not meet these criteria. Note
that these criteria allow IPv4 addresses in dotted-decimal or other notations,
e.g. {{?RFC3986, Section 7.4}} and {{WHATWG-IPV4}}. Clients SHOULD ignore the
`ECHConfig` if it contains an encoded IPv4 address. (To determine if a
public_name value is an IPv4 address, clients can invoke the IPv4 parser
algorithm in {{WHATWG-IPV4}}. It returns a value when the input is an IPv4
address.) See {{auth-public-name}} for how the client interprets and validates

> the public_name.

Reviewer: Try splitting this up a little.  You can start multiple sections with ": " to make separate paragraphs (I think).

Author: Oh, nice. I didn't know that was possible. I broke up the description into logical units.

the server uses wildcard names or added new names to the anonymity set.

public_name
: The DNS name of the client-facing server, i.e., the entity trusted
to update the ECH configuration. This is used to correct misconfigured clients,

> as described in {{handle-server-response}}.

Reviewer 2: Let's separate out the conformance requirements and the check. I think you want to say "This value MUST be a valid domain name in accordance with XXX".  I believe that would still allow trailing (but not leading) dot, so you may also have to say "The value MUST NOT end with an ASCII dot".

Author: Good call -- I believe your suggestions rephrased things to make the requirements and check read easier.

> : This value MUST NOT begin or end with an ASCII dot and MUST be parsable as a
> dot-separated sequence of LDH labels, as defined in
> {{!RFC5890, Section 2.3.1}}. Clients MUST ignore any `ECHConfig` structure
> whose public_name does not meet these criteria.
> 

Reviewer 2: ```suggestion
: Clients MUST ignore any `ECHConfig` structure
whose public_name does is not parsable as a
dot-separated sequence of LDH labels, as defined in
{{!RFC5890, Section 2.3.1}} or which
begins or end with an ASCII dot.
```

Reviewer 2: Note that I think this still allows for bogus names, like ones that are super-long. Do we want to allow clients to check for that?

Author: The public_name field is already limited to 255 characters, so I think we're fine.

> : These criteria allow IPv4 addresses in dotted-decimal or other notations,
> e.g. {{?RFC3986, Section 7.4}} and {{WHATWG-IPV4}}. Clients SHOULD ignore the
> `ECHConfig` if it contains an encoded IPv4 address. To determine if a
> public_name value is an IPv4 address, clients can invoke the IPv4 parser
> algorithm in {{WHATWG-IPV4}}. It returns a value when the input is an IPv4
> address.

Reviewer 2: ```suggestion
:  Clients SHOULD ignore the
`ECHConfig` if it contains an encoded IPv4 address. To determine if a
public_name value is an IPv4 address, clients can invoke the IPv4 parser
algorithm in {{WHATWG-IPV4}}. It returns a value when the input is an IPv4
address.
```
